WIP: allow debugging of the original C# source

danwalmsley opened the pull request:
- adds #line directives based on the SequencePoint for the instruction.
- squishes statements onto 1 line to prevent clang / gcc incrementing the line number (this is hack and could be handled better).

sample code:
```
// See https://aka.ms/new-console-template for more information

using System;
using Dntc.Attributes;

namespace HelloWorld;

static class Program
{
    [NativeFunctionCall("printf", "<stdio.h>")]
    public static void Printf(string value)
    {
    }
    
    [NativeFunctionCall("printf", "<stdio.h>")]
    public static void Printf<T>(string template, T input1)
    {
        
    }
    
    [CustomFunctionName("main")]
    public static int Main()
    {
        for (int i = 0; i < 1000; i++)
        {
            Printf("%u Hello, World!" + "\n", i);
        }

        return 0;
    } 
}
```
sample code generation:
```
#include <stdint.h>
#include <stdbool.h>
#include <stdio.h>
#include "HelloWorld.h"


int32_t main(void) {

#line 23 "/Users/dan/repos/dntc/Samples/HelloWorld/Program.cs" // 23:23,5:6
	int32_t __local_0 = {0};	bool __local_1 = {0};	int32_t __local_2 = {0};
#line 24 "/Users/dan/repos/dntc/Samples/HelloWorld/Program.cs" // 24:24,14:23
	__local_0 = 0;	goto main_IL_0017;
#line 26 "/Users/dan/repos/dntc/Samples/HelloWorld/Program.cs" // 26:26,13:50
main_IL_0005:	printf("%u Hello, World!\n", __local_0);
#line 24 "/Users/dan/repos/dntc/Samples/HelloWorld/Program.cs" // 24:24,35:38
	__local_0 = (__local_0 + 1);
#line 24 "/Users/dan/repos/dntc/Samples/HelloWorld/Program.cs" // 24:24,25:33
main_IL_0017:	__local_1 = (__local_0 < 1000);	if (__local_1) {		goto main_IL_0005;	}
#line 29 "/Users/dan/repos/dntc/Samples/HelloWorld/Program.cs" // 29:29,9:18
	__local_2 = 0;
#line 29 "/Users/dan/repos/dntc/Samples/HelloWorld/Program.cs" // 29:29,9:18
	goto main_IL_0027;
#line 30 "/Users/dan/repos/dntc/Samples/HelloWorld/Program.cs" // 30:30,5:6
main_IL_0027:	return __local_2;}
```


https://github.com/user-attachments/assets/88859393-f424-46db-863f-d784232162ae

danwalmsley commented:
@KallDrexx firstly awesome work here. What do you think about having debug info mapping like this,  I read about another transpiler using #line directives to achieve this, is this the right approach?

danwalmsley commented:
btw the comments at the end of the line directives ( // 30:30,5:6) these are StartLine:EndLine,StartColumn:EndColumn) this extra information could somehow be given to the debugger but not sure how….

KallDrexx commented:
This is awesome!  I kept meaning to add something to help the correlation between the C side and the C# side but didn't make it this last step so I 1000% appreciate this!  I haven't looked fully at the code yet (packing for a weekend trip) but I can do a better look this evening.  

One thing to note (that's not well documented and I keep meaning to add CI steps for it) is the sample probably isn't necessary.  If you run `Samples/build_all.sh` it will build all the sample projects and the generated C code will be checked into source control.  

Due to time constraints I don't have a formal automated testing system in place to validate changes, so what I use right now are the existing sample projects as the tests.  Packaging the generated code with the commit that changed them allows me and other potential users to understand what commits caused a specific code generation look to better isolate bugs or find regressions.  

My plan is to eventually have CI block PRs that have generated code changes not in the changeset.   

The only *immediate* comment I have here is using the `#line` directive.  I actually didn't know this was a thing but reading up on it that's a better idea than I had.  My only concern is how well supported that directive is (for instance, I don't know if that would work for the SNES C compiler)..  We might want to plumb an option to swap between
1. `#line ` directives
2. Comments
3. Bypass

That way transpiler plugins can customize how they result for maximum capacity.

KallDrexx commented:
I just watched your video and yesssss, I didn't put it together that it allows debugger hooks doing this.  That's amazing.

danwalmsley commented:
Did a nice refactor:

1. added IStatementGenerator… where you can hook into before and after statements are generated… this means the generation of the line info is no longer hardcoded into the transpiler.. its a plugin if thats the correct wording.

2. implemented a LineInfoStatementGenerator, that inserts the lineinfo, and changes streamwriter.Newline to and empty string, to make sure the following statements will be bunched onto a single line.

It then adds end EndLineStatement which will retore the newline behaviour.
it also removes duplicate line info where the line info is the same.

3. added a property to the manifest so the user can select LineInfoMode = “None/Directives/Comments”

I guess we could have a Mutator? that can mutate or format how the line info should look? not sure where that should live?


Mode None:
```
#include <stdint.h>
#include <stdio.h>
#include <stdbool.h>
#include "HelloWorld.h"


int32_t main(void) {
	int32_t i = {0};
	i = 0;
	goto main_IL_0013;

main_IL_0004:
	printf("%u Hello, World!\n", i);
	i = (i + 1);

main_IL_0013:
	if ((i < 1000)) {
		goto main_IL_0004;
	}
	return 0;
}
```

Mode Comments:
```
#include <stdint.h>
#include <stdio.h>
#include <stdbool.h>
#include "HelloWorld.h"


int32_t main(void) {
// "/Users/dan/repos/dntc/Samples/HelloWorld/Program.cs" [23 14 - 23 23]
	int32_t i = {0};	i = 0;	goto main_IL_0013;
// "/Users/dan/repos/dntc/Samples/HelloWorld/Program.cs" [25 13 - 25 50]
main_IL_0004:	printf("%u Hello, World!\n", i);
// "/Users/dan/repos/dntc/Samples/HelloWorld/Program.cs" [23 35 - 23 38]
	i = (i + 1);
// "/Users/dan/repos/dntc/Samples/HelloWorld/Program.cs" [23 25 - 23 33]
main_IL_0013:	if ((i < 1000)) {		goto main_IL_0004;	}
// "/Users/dan/repos/dntc/Samples/HelloWorld/Program.cs" [28 9 - 28 18]
	return 0;
}
```

Mode Directives:
```
#include <stdint.h>
#include <stdio.h>
#include <stdbool.h>
#include "HelloWorld.h"


int32_t main(void) {
#line 23 "/Users/dan/repos/dntc/Samples/HelloWorld/Program.cs" // [23 14 - 23 23]
	int32_t i = {0};	i = 0;	goto main_IL_0013;
#line 25 "/Users/dan/repos/dntc/Samples/HelloWorld/Program.cs" // [25 13 - 25 50]
main_IL_0004:	printf("%u Hello, World!\n", i);
#line 23 "/Users/dan/repos/dntc/Samples/HelloWorld/Program.cs" // [23 35 - 23 38]
	i = (i + 1);
#line 23 "/Users/dan/repos/dntc/Samples/HelloWorld/Program.cs" // [23 25 - 23 33]
main_IL_0013:	if ((i < 1000)) {		goto main_IL_0004;	}
#line 28 "/Users/dan/repos/dntc/Samples/HelloWorld/Program.cs" // [28 9 - 28 18]
	return 0;
}
```

does the IStatementGenerator stuff make sense in the “Statements/Generators” folder?

danwalmsley commented:
Another thing iv added to make the debugging better is to maintain the local variable names where possible.

KallDrexx commented on Dntc.Common/Conversion/PlannedFileConverter.cs:
```diff
@@ -107,18 +112,31 @@ private IReadOnlyList<CStatementSet> GetMethodStatements(
         
+        var methodInstruction = dotNetDefinedMethod.Definition.Body.Instructions.FirstOrDefault();
+
+        if (methodInstruction != null)
```
Can this actually be null?  Won't the foreach on 140 crash if that's the case?

danwalmsley replied:
fixed

KallDrexx commented on Dntc.Common/Syntax/MethodBlock.cs:
```diff
@@ -115,6 +126,7 @@ private static IReadOnlyList<CStatementSet> MoveDeclarationsToBeginning(IReadOnl
 
-        return declarations.Concat(nonDeclarations).ToArray();
+        // Combine in the order: initialLineStatements -> declarations -> other non-declarations
+        return initialLineStatements.Concat(declarations).Concat(nonDeclarations).ToArray();
```
Is it worth keeping all the line statements?  I assume only the last one would be useful.

danwalmsley replied:
fixed

KallDrexx commented on Dntc.Common/Syntax/Statements/LineStatementSet.cs:
```diff
@@ -0,0 +1,26 @@
+        switch (Mode)
+        {
+            case LineInfoMode.Directives:
+                await writer.WriteLineAsync($"#line {SequencePoint.StartLine} \"{SequencePoint.Document.Url}\" // [{SequencePoint.StartLine} {SequencePoint.StartColumn} - {SequencePoint.EndLine} {SequencePoint.EndColumn}]");
```
Is it worth prefacing both of these modes with a `\n` at the beginning of the line?  

That might help keep the C code a bit more readable so instead of 

```c
int32_t main(void) {
// "/Users/dan/repos/dntc/Samples/HelloWorld/Program.cs" [23 14 - 23 23]
	int32_t i = {0};	i = 0;	goto main_IL_0013;
// "/Users/dan/repos/dntc/Samples/HelloWorld/Program.cs" [25 13 - 25 50]
main_IL_0004:	printf("%u Hello, World!\n", i);
```

it generates as
```c
int32_t main(void) {

// "/Users/dan/repos/dntc/Samples/HelloWorld/Program.cs" [23 14 - 23 23]
	int32_t i = {0};	i = 0;	goto main_IL_0013;

// "/Users/dan/repos/dntc/Samples/HelloWorld/Program.cs" [25 13 - 25 50]
main_IL_0004:	printf("%u Hello, World!\n", i);
```

There are times that I've needed to double check logic in C, and even though it's not the most readable thing this might be helpful?

The alternative is the `EndLineStatements` which I kind of thought would be what accomplishes that but the example you put in the latter PR comments didn't show it (maybe that was added after?)

danwalmsley replied:
iv made it put the spacing in, much more readable, I used the EndLineStatements to do it.

KallDrexx commented on Dntc.Common/Syntax/Statements/Generators/LineInfoStatementGenerator.cs:
```diff
@@ -0,0 +1,71 @@
+                var lastEndLineStatement = statements.OfType<EndLineStatementSet>().LastOrDefault();
+                if (lastEndLineStatement != null)
+                {
+                    statements.Remove(lastEndLineStatement);
```
Am I correct that this is handling the case where multiple line statement generators are in play for the same section, so we merge their line statements together?

And what's the reason for the last end line statement from line 43 to be assigned to `_movedstatementSet` but not this one?

danwalmsley replied:
yes thats correct, merging where sequence points are the same.
Iv added some comments to make this more understandable.

The moveStatementSet is for handling hidden sequence points, they essentially become part of the same line, so we just keep moving the endLineStatement until there is a non hidden sequence point.

KallDrexx reviewed (commented):
This is great and much appreciated.  Especially happy to have named locals when available, I didn't even think to see if that was available.

Just have a few small questions, and running `Samples/build_all.sh` and checking the generated code in will be required for a PR to be merged.  

Also any comments on how easy to learn the code base is and any pain points felt would be appreciated!

KallDrexx commented:
Also, can you explain why directives require the statements to be on a single line

danwalmsley commented:
> Also, can you explain why directives require the statements to be on a single line

I got that clue from here:
https://www.reddit.com/r/ProgrammingLanguages/comments/ydyw32/comment/ivcs1jq/?utm_source=share&utm_medium=web3x&utm_name=web3xcss&utm_term=1&utm_content=share_button

https://www.reddit.com/r/ProgrammingLanguages/comments/ydyw32/comment/iveb4mb/?utm_source=share&utm_medium=web3x&utm_name=web3xcss&utm_term=1&utm_content=share_button

basically what happens is the C compiler ends up incrementing the line number itself… for each line in C, so then the debug information doesnt match the sequence points in the C# code.

I think the alternative to removing the newlines is to put a line directive before every single line.

however if you need to debug the generated code you can turn this off, and you get quite nicely formatted C without it all being bunched onto a single line.

KallDrexx commented:
That makes perfect sense and I agree with you then.  I may make a command line switch to override the manifest's line mode but I can do that later.  

I just realized the samples will be impossible to pass CI checks unless they are marked as `LineMode.None`, since the line directives are absolute pathed.  So if you make that change I assume the samples should have no code gen changes and it would pass.  I'd be good with merging it after that.

danwalmsley commented:
@KallDrexx perhaps instead of LineInfoMode we could call it “DebugInfoMode” the options could be “C”, “CSharpDirectives”, “CSharpComments” or is it more understandable as it is?

danwalmsley commented:
@KallDrexx I think I have addressed the issues you raised, there build is failing because of the preservation of variable names.

if you want to check it and are happy should I then commit the generated code changes to make the CI pass?

KallDrexx commented:
`DebugInfoMode` is a great name actually.  Maybe `None`, `CLineSourceMaps` and `CSharpLineComments` (the last one I'm a bit more iffy on).

Everything looks good to me except for one minor thing.  The octahedron sample manifests in https://github.com/KallDrexx/dntc/tree/main/Samples/Octahedron will need to have the mode set to `None` for the CI tests to pass (plus having the code generation checked into the commit.

Once that's done I'm extremely happy to merge this!

danwalmsley commented:
@KallDrexx added a fix, to use invariantculture when generating numeric literals (3.14 became 3,14 on my machine).

ok iv just checked in the generated code and made the last changes you suggested.

hopefully ready now

KallDrexx commented:
Hah I guess the problem with using C# variable names is the differences in scopes.  

I wonder if we should have the naming be something like `__local_0_x` instead of just `x` to deduplicate it?

danwalmsley commented:
"Also any comments on how easy to learn the code base is and any pain points felt would be appreciated!”

So far its been fairly straightforward :)

danwalmsley commented:
> Hah I guess the problem with using C# variable names is the differences in scopes.
> 
> I wonder if we should have the naming be something like `__local_0_x` instead of just `x` to deduplicate it?

Pushed a fix to normalize the list of declarations… luckily you had that array test function its the only place that reproduces the issue.

KallDrexx commented:
Awesome.  Very very very much appreciated!  I forgot to rename the commit to not have WIP before pressing the button but on well.
